Source/SwiftLintFramework/Models/LinterCache.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  LinterCache.swift
+//  SwiftLint
+//
+//  Created by Marcelo Fabri on 12/27/16.
+//  Copyright © 2016 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+public enum LinterCacheError: Error {
+    case invalidFormat
+    case differentVersion
+    case differentConfiguration
+}
+
+public final class LinterCache {
+    private var cache: [String: Any]
+    private let lock = NSLock()
+
+    public init(currentVersion: Version = .current, configurationHash: Int? = nil) {
+        cache = [String: Any]()
+        cache["version"] = currentVersion.value
+        cache["configuration_hash"] = configurationHash
+        cache["files"] = [:]
+    }
+
+    public init(cache: Any, currentVersion: Version = .current, configurationHash: Int? = nil) throws {
+        guard let dictionary = cache as? [String: Any] else {
+            throw LinterCacheError.invalidFormat
+        }
+
+        guard let version = dictionary["version"] as? String, version == currentVersion.value else {
+            throw LinterCacheError.differentVersion
+        }
+
+        if dictionary["configuration_hash"] as? Int != configurationHash {
+            throw LinterCacheError.differentConfiguration
+        }
+
+        self.cache = dictionary
+    }
+
+    public convenience init(contentsOf url: URL, currentVersion: Version = .current,
+                            configurationHash: Int? = nil) throws {
+        let data = try Data(contentsOf: url)
+        let json = try JSONSerialization.jsonObject(with: data, options: [])
+        try self.init(cache: json, currentVersion: currentVersion,
+                      configurationHash: configurationHash)
+    }
+
+    public func cacheFile(_ file: String, violations: [StyleViolation], hash: Int) {
```

Comment:
Doesn't conform to Swift 3 API Design Guidelines, and implementation could be simplified:

```swift
public func cache(violations: [StyleViolation], forFile file: String, fileHash: Int) {
    lock.lock()
    var filesCache = (cache["files"] as? [String: Any]) ?? [:]
    filesCache[file] = [
        "violations": violations.map(dictionaryForViolation),
        "hash": fileHash
    ]
    cache["files"] = filesCache
    lock.unlock()
}
```

Adds an insignificant amount of work inside the lock.